### PR TITLE
Add error type and resource exhausted cause to task error metrics

### PR DESCRIPTION
**Pull request by rodrigozhou**
## What changed?

Add error type and resource exhausted cause to task error metrics

## Why?

Be able to identify the error type and resource exhausted cause in task errors.

### CODE REVIEW

File: service/history/queues/executable.go [hunk 1]
```diff
@@ -494,6 +494,14 @@ func (e *executableImpl) HandleErr(err error) (retErr error) {
 		}
 	}()
 
+	metricsHandler := e.taggedMetricsHandler.WithTags(metrics.ServiceErrorTypeTag(err))
```
Comment (bergundy) [inline service/history/queues/executable.go:497]: Make sure this metric always has the same number of tags. Prometheus complains about variable tags. Also need to check what the process is to add tags to existing metrics.

**Review by bergundy — COMMENTED** [review]
Can you please double check that we will always emit this extra tag for these metrics now?

Also would like @yycptt and @prathyushpv to approve.

File: service/history/queues/executable.go [hunk 2]
```diff
@@ -517,21 +529,22 @@ func (e *executableImpl) HandleErr(err error) (retErr error) {
...
 		return rewrittenErr
 	}
+
 	// Unexpected errors handled below
 	e.unexpectedErrorAttempts++
-	metrics.TaskFailures.With(e.taggedMetricsHandler).Record(1)
+	metrics.TaskFailures.With(metricsHandler).Record(1)
 	e.logger.Warn("Fail to process task", tag.Error(err), tag.ErrorType(err), tag.UnexpectedErrorAttempts(int32(e.unexpectedErrorAttempts)), tag.LifeCycleProcessingFailed)
 
 	if e.isUnexpectedNonRetryableError(err) {
```
Comment (yycptt) [inline service/history/queues/executable.go:538]: we can remove TaskInternalErrorCounter inside this func as well, now that we can always get this info from TaskFailures metric.

File: service/history/queues/executable.go [hunk 3]
```diff
@@ -494,6 +494,18 @@ func (e *executableImpl) HandleErr(err error) (retErr error) {
...
+	var resourceExhaustedErr *serviceerror.ResourceExhausted
+	var resourceExhaustedCauseTag metrics.Tag
+	if errors.As(err, &resourceExhaustedErr) {
+		resourceExhaustedCauseTag = metrics.ResourceExhaustedCauseTag(resourceExhaustedErr.Cause)
+	} else {
+		resourceExhaustedCauseTag = metrics.ResourceExhaustedCauseEmptyTag()
+	}
+	metricsHandler := e.taggedMetricsHandler.WithTags(
+		metrics.ServiceErrorTypeTag(err),
+		resourceExhaustedCauseTag,
```
[the remark this answers is not in the document]
Comment (rodrigozhou) [reply service/history/queues/executable.go]: Do you mean to add the busy workflow counter to TaskThrottledCounter with the resource exhausted cause?

File: service/history/queues/executable.go (hunk 2 above)
Comment (rodrigozhou) [reply service/history/queues/executable.go]: TaskFailures doesn't have any tags with the metrics.

File: service/history/queues/executable.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (yycptt) [reply service/history/queues/executable.go]: yeah I mean we don't need a separate metric counter for the wf busy case. It's the same thing as TaskThrottled counter, which cause equals to workflow busy.

File: service/history/queues/executable.go (hunk 2 above)
Comment (yycptt) [reply service/history/queues/executable.go]: I thought you want to add error type tag to TaskFailures metric?

Comment (rodrigozhou) [reply service/history/queues/executable.go]: But if I only add the cause tag to TaskThrottledCounter, I don't need to add the error type to TaskFailures.

**Review by yycptt — APPROVED** [review]

File: service/history/queues/executable.go (hunk 2 above)
Comment (yycptt) [reply service/history/queues/executable.go]: Sure. I am asking because from the PR title and desc, it seems like you want both, hence the question.

**Pull request merged**